semgrep-core/src/engine/Dataflow_tainting.ml

```diff
@@ -184,28 +185,49 @@ let report_findings env findings =
   if findings <> [] then
     env.config.handle_findings env.fun_name findings env.var_env
 
-let unify_meta_envs env1 env2 =
+let unify_mvars_sets mvars1 mvars2 =
   let xs =
     List.fold_left
       (fun xs (mvar, mval) ->
         let* xs = xs in
-        match List.assoc_opt mvar env2 with
+        match List.assoc_opt mvar mvars2 with
         | None -> Some ((mvar, mval) :: xs)
         | Some mval' ->
             if Metavariable.equal_mvalue mval mval' then
               Some ((mvar, mval) :: xs)
             else None)
-      (Some []) env1
+      (Some []) mvars1
   in
   let ys =
-    List.filter (fun (mvar, _) -> not @@ List.mem_assoc mvar env1) env2
+    List.filter (fun (mvar, _) -> not @@ List.mem_assoc mvar mvars1) mvars2
   in
   Option.map (fun xs -> xs @ ys) xs
 
+let sink_biased_union_mvars source_mvars sink_mvars =
+  let source_mvars' =
+    List.filter
+      (fun (mvar, _) -> not @@ List.mem_assoc mvar sink_mvars)
+      source_mvars
+  in
+  Some (source_mvars' @ sink_mvars)
+
+let merge_source_sink_mvars env =
```

Comment:
The names of these functions are a little too compact to serve as descriptions. I think if I didn't already have context, even with the comments already here, I would be a bit confused about what's going on. Would be nice to have a function-level comment stating that we're merging the metavariables bound in the source pattern and the sink pattern.

Also, what happens to metavariables in sanitizers?

Reply:
I believe we were also meant to unify source-sanitizer metavariables, but it seems that we never did. I added a `TODO` comment in a previous PR.